Expanded Conversation Attachments

DeveloperBlue opened the pull request:
Hope everyone's doing well, this is my first flutter PR.

This PR allows you to:
- See all local attachments for a conversation
- Filter attachments by sender, type, date, and search
- Jump to the attachment message in the conversation

### Viewing More Attachments
https://github.com/user-attachments/assets/845a2dcf-5174-48f2-9fcc-8ddf03cbc475

### Filtering Attachments
https://github.com/user-attachments/assets/10b27075-93af-451e-8f5d-ad82a43a223e

https://github.com/user-attachments/assets/d7c9f025-bbd7-4674-8602-7c10848f657d

### Attachment Popups
https://github.com/user-attachments/assets/e1c281fd-d3e3-40bc-b201-b4618b3896c7

### Jump to Message
https://github.com/user-attachments/assets/089d198a-5cbf-4a50-983d-c1c409c1114d

## Additional Information
- Advanced filtering allows you to search for a specific sender in group chats instead of just the general "From Others".
- When the "High Performance" setting is enabled, the search bar switches from live search (searching on every character) to presenting the user with a manual submission button.
- Adds a small "From (CONTACT)" button/label to the media viewer. This button also lets you jump to the message location in the conversation.
- The Attachments Popup widgets are taken from the Message Popup widgets. They were heavily modified but do include some duplicate logic.
- Had to add an "isEquals" checker method to the Handle object, because just comparing two objects (==) fails when their "fake" names under redacted mode are different.
- Respects ordering the popup context menu's options based on user's "Message Options Order" settings. 
-  Only the local attachments are shown. It does not fetch any attachments from the server. You can work around this by continually hitting the "Fetch More Messages" button on the Conversation Details page. If anyone understands the custom scroll views better, it could be useful to have it automatically pull the next 100 attachments when the user reaches the bottom of the scroll view.

## Issues
- ~~[ ] ⚠ Sometimes the "Jump To Message" action gets stuck in an infinite pending state. Exiting the conversation and jumping again seems to fix this issue. **This should be resolved before the PR is merged, and I need a little help figuring out why it does this.** [Area 1](https://github.com/BlueBubblesApp/bluebubbles-app/blob/1c39563bf6b4ae038457d059718ab3c04277585d/lib/app/layouts/fullscreen_media/fullscreen_holder.dart#L156-L167), [Area 2](https://github.com/BlueBubblesApp/bluebubbles-app/blob/1c39563bf6b4ae038457d059718ab3c04277585d/lib/app/layouts/conversation_attachments/widgets/attachment_popup.dart#L557-L570)~~

And some more extensive testing is required, since I was only able to test on my android emulator. I have yet to test the changes on the web or desktop builds. I'm @DeveloperBlue on the Discord as well.

DeveloperBlue committed:
Add personal developer build flavor

DeveloperBlue committed:
Add isEqual helper checker to Handle models

DeveloperBlue committed:
Add ContactAvatarWidget button to iOS Fullscreen Holder View

DeveloperBlue committed:
Add JumpToMessage action to DetailsMenuAction

DeveloperBlue committed:
Add 'See More' header and navigator to Conversation Details view

DeveloperBlue committed:
Add Attachments Filter bottom sheet

DeveloperBlue committed:
Add feat: Expanded Conversation Details View

DeveloperBlue committed:
Tweak attachment popup options sort order and numberToShow

DeveloperBlue committed:
Tweak default attachment popup options sort order

DeveloperBlue committed:
bugfix: Undo adjustments to Details Menu Action options

DeveloperBlue committed:
bugfix: Hide JumpToMessage button when attachment has not been sent yet

zlshames commented:
An awesome addition to the app! The only change I'll request is that you center the filters in the popup window, to be consistent with other popup UIs (i.e. the search filtering).

I have a couple of questions as well:

> Adds a small "From (CONTACT)" button/label to the media viewer. This button also lets you jump to the message location in the conversation.

Does this add it to all media viewer pages or just the one in the chat details?

> The Attachments Popup widgets are taken from the Message Popup widgets. They were heavily modified but do include some duplicate logic.

If we make changes to the other popups (UI-wise), I assume we'll also need to also update the new message popup holder?

> Sometimes the "Jump To Message" action gets stuck in an infinite pending state. Exiting the conversation and jumping again seems to fix this issue. This should be resolved before the PR is merged, and I need a little help figuring out why it does this. [Area 1](https://github.com/BlueBubblesApp/bluebubbles-app/blob/1c39563bf6b4ae038457d059718ab3c04277585d/lib/app/layouts/fullscreen_media/fullscreen_holder.dart#L156-L167), [Area 2](https://github.com/BlueBubblesApp/bluebubbles-app/blob/1c39563bf6b4ae038457d059718ab3c04277585d/lib/app/layouts/conversation_attachments/widgets/attachment_popup.dart#L557-L570)

Our jump to message code is sometimes bugged, which is why some people get freezes/crashes when certain chats have it enabled. I don't have a fix at this time, so I think maybe removing it as a feature until we figure out how to better do it.

Otherwise... it looks great! Thank you for your contribution!

zlshames reviewed (changes requested):
See my comments above

DeveloperBlue committed:
Center filter options on slideup panel

DeveloperBlue committed:
Remove/hide Jump to Message feature

DeveloperBlue commented:
> The only change I'll request is that you center the filters in the popup window, to be consistent with other popup UIs (i.e. the search filtering).

Is this what you mean by centering the filters? I've gone ahead and horizontally centered them with the last push, but I see that the filter for searching conversations is left-aligned on my other devices.

I've also removed/hid any UI for jumping to the attachment's message for now.

<p float="left">
  <img src="https://github.com/user-attachments/assets/c4f918be-8107-4a79-95df-a2315aa86254" height="360" />
  <img src="https://github.com/user-attachments/assets/48ac9aa6-73c0-4e5b-8184-e064d3c09993" height="360" /> 
</p>

----

> Does this add it to all media viewer pages or just the one in the chat details?

It is added to all media viewer pages. (Or more specifically, it's added to the existing media viewer page that all images/videos navigate to when you tap them.) **Since I removed the Jump to Message feature for now, the button/label is now gone has been reverted to how it was before this PR.** 

> If we make changes to the other popups (UI-wise), I assume we'll also need to also update the new message popup holder?

Yes. In the future, I think a global popup component could be designed and then extended from for both components. I wouldn't mind looking into this.

Please let me know if you have any more feedback, questions, or changes! Thanks again

DeveloperBlue committed:
Minor code cleanup / touch less files when merging

DeveloperBlue committed:
bugfix: fix redownload attachment feature

zlshames commented:
> but I see that the filter for searching conversations is left-aligned on my other devices.

You're right... i gotta fix that then ;)

> I've also removed/hid any UI for jumping to the attachment's message for now.

Perfect, thank ya

> Yes. In the future, I think a global popup component could be designed and then extended from for both components. I wouldn't mind looking into this.

Sounds good, thank you for the contribution! I'll check it out after we do this Desktop release (happening shortly). Thanks Joel! 

Then we'll have to take a crack at the next Android update, which we'll include this in it.

P.S. We know that we need to do a better job of making our components better and more "static" in a way. Less code within them and focus on minimizing widget tree updates to improve performance. Only issue is that it's a fairly large effort to take on with how complex the code base is. We also want to improve our chat manager services and what not to better optimize how the widget tree is updated.

zlshames commented on android/app/build.gradle:
```diff
@@ -85,6 +85,13 @@ android {
             resValue "string", "file_provider", "com.bluebubbles.messaging.tanay.fileprovider"
             applicationId "com.bluebubbles.messaging.tanay"
         }
+        developerblue {
```
Can you remove this please :)

zlshames commented on android/app/build.gradle:
```diff
@@ -113,6 +120,7 @@ android {
         release {
             productFlavors.joel.signingConfig signingConfigs.debug
             productFlavors.tanay.signingConfig signingConfigs.debug
+            productFlavors.developerblue.signingConfig signingConfigs.debug
```
Please remove

DeveloperBlue committed:
Remove personal build flavor

DeveloperBlue commented:
Resolved the changes to build.gradle 👍

zlshames commented:
Sorry, been slackin, i plan on reviewing these soon

zlshames commented:
A few things to request when you rewrite this for the dev branch:

1. I've updated the UI for the attachment cards, so i'd like your changes to be consistent with those.
2. I've updated the full screen viewer for images & videos. It should just be plug n play, but just make sure it follows the new design elements.
